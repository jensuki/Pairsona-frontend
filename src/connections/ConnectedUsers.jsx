import { useState, useEffect, useContext } from 'react';
import { useNavigate, Link } from 'react-router-dom';
import Api from '../helpers/api';
import UserContext from '../context/UserContext';
import getProfilePic from '../helpers/getProfilePic';
import Loading from '../components/Loading';
import noConnections from '../assets/placeholders/no_connections.png';
import '../styles/Connections.css';

/** component for displaying list of all connected users */

const ConnectedUsers = () => {
    const { currUser, hasNewMessages, checkNotifications, loading } = useContext(UserContext);
    const navigate = useNavigate();

    const [connections, setConnections] = useState(null);
    const [error, setError] = useState(null);

    // fetch connections whenever curr user or notifications change
    useEffect(() => {
        if (!currUser) return;

        const fetchConnections = async () => {
            try {
                const res = await Api.getConnections();
                setConnections(res || []); // update state with response
                checkNotifications();
            } catch (err) {
                setError(err.message);
            }
        };

        fetchConnections();
    }, [currUser, checkNotifications]);


    // handle removing a connection
    const handleRemoveConnection = async (connectionId) => {
        try {
            // proceed with removal
            await Api.removeConnection(connectionId);

            // update connection list after removal
            setConnections((prevConnections) =>
                prevConnections.filter(conn => conn.connectionId !== connectionId)
            );
            // refresh notifications immediately
            checkNotifications();
        } catch (err) {
            setError(err.message);
        }
    };

    // handle navigating to messaging page
    const handleSendMessage = (username) => {
        navigate(`/messages/${username}`);
    };

    if (error) {
        return <p className="error">{error}</p>;
    }

    if (loading || connections === null) {
        return <Loading />;
    }

    return (
        <div className="ConnectedUsers">
            <h2>Your Connections</h2>

            {connections.length === 0 && <img src={noConnections} />}

            <ul className="connections-list">
                {connections.map(conn => {
                    return (
                        <li key={conn.connectionId} className="connection-item">
                            {/* connection list */}
                            <div className="connection-info">
                                <img
                                    src={getProfilePic(conn.profilePic)}
                                    alt={`${conn.firstName} ${conn.lastName}`}
                                    className="profile-pic"
                                    loading="lazy"
                                />
                                <div className="user-details">
                                    <Link to={`/profile/${conn.username}`} className="user-name">
                                        {conn.firstName} {conn.lastName}
                                    </Link>
                                    {hasNewMessages.includes(conn.username.toLowerCase()) && (
                                        <span className="notification-dot"></span>
                                    )}
                                </div>
                            </div>
                            <div className="connection-actions">
                                <button className="chat-btn" onClick={() => handleSendMessage(conn.username)}>
                                    Chat with {conn.firstName}
                                </button>
                                <button className="remove-btn" onClick={() => handleRemoveConnection(conn.connectionId)}>
                                    Remove
                                </button>
                            </div>
                        </li>
                    );
                })}
            </ul>
        </div>
    );
};

export default ConnectedUsers;
