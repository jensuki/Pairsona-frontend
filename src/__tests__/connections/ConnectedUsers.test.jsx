import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { UserProvider } from '../../test_common/setupTests';
import ConnectedUsers from '../../connections/ConnectedUsers';
import { expect, describe, test, vi, beforeEach } from 'vitest';
import Api from '../../helpers/api';
import { demoConnections, demoNotifications, demoUser } from '../../test_common/mockData';

// mock api calls
vi.mock('../../helpers/api', () => ({
    default: {
        getConnections: vi.fn(),
        removeConnection: vi.fn(),
    },
}));

// mock useNavigate
const mockNavigate = vi.fn();

vi.mock('react-router-dom', async (importActual) => {
    const actual = await importActual(); // import actual module
    return {
        ...actual, // keep original exports
        useNavigate: () => mockNavigate
    }
})

// helper
const renderConnectedUsers = () => {
    return render(
        <MemoryRouter>
            <UserProvider
                currUser={demoUser}>
                <ConnectedUsers />
            </UserProvider>
        </MemoryRouter >
    );
};

describe('ConnectedUsers component', () => {
    beforeEach(() => {
        vi.clearAllMocks();

        // common mock setup for connected users
        Api.getConnections.mockResolvedValue(demoConnections);
    });

    test('renders without crashing', async () => {
        renderConnectedUsers();

        await waitFor(() => {
            expect(screen.getByText(/Your Connections/i)).toBeInTheDocument();
        });

        // verify user3 connection is rendered
        expect(screen.getByText(/User\s+Three/i)).toBeInTheDocument();
    });

    test('matches snapshot', async () => {
        const { asFragment } = renderConnectedUsers();

        await waitFor(() => {
            expect(screen.getByText(/Your Connections/i)).toBeInTheDocument();
        });

        // check if the mock is called
        expect(Api.getConnections).toHaveBeenCalled();

        // verify snapshot
        expect(asFragment()).toMatchSnapshot();
    });

    test('displays list of connections', async () => {
        renderConnectedUsers();

        // verify connection details are rendered
        await waitFor(() => {
            demoConnections.forEach((conn) => {
                expect(screen.getByText(`${conn.firstName} ${conn.lastName}`)).toBeInTheDocument();
            })
        })
    })

    test('removes a connection when "remove" button is clicked', async () => {
        // mock successful api repsonse for removal
        Api.removeConnection.mockResolvedValue();

        renderConnectedUsers();

        // wait for connections to render
        await waitFor(() => {
            expect(screen.getByText(/User\s+Three/i)).toBeInTheDocument();
        })

        // click remove
        const removeBtn = screen.getByText(/Remove/i);
        fireEvent.click(removeBtn);

        // ensure api call was made for removing connection set in mock data
        expect(Api.removeConnection).toHaveBeenCalledWith(demoConnections[0].connectionId);

        // ensure no more connection
        await waitFor(() => {
            expect(screen.queryByText(/User\s+Three/i)).not.toBeInTheDocument();
        })
    })

    test('navigates to chat page on "Chat with {username}" click', async () => {
        renderConnectedUsers();

        // verify the chat btn is rendered
        await waitFor(() => {
            expect(screen.getByText(/Chat with User/i)).toBeInTheDocument();
        })

        // click chat btn
        const chatBtn = screen.getByText(/Chat With User/i);
        fireEvent.click(chatBtn);

        // verify proper navigation to chat page with user3 through custom navigate hook
        expect(mockNavigate).toHaveBeenCalledWith(`/messages/${demoConnections[0].username}`)

    })

    test('displays notification dot for users with new messages', async () => {
        render(
            <MemoryRouter>
                <UserProvider
                    currUser={demoUser}
                    hasNewMessages={demoNotifications.hasNewMessages}>
                    <ConnectedUsers />
                </UserProvider>
            </MemoryRouter>
        )
        // wait for connections to render
        await waitFor(() => {
            expect(screen.getByText(/User Three/i)).toBeInTheDocument();
        })

        // check if notifcation dot exists for user3
        const notificationDot = screen.getByText(/User Three/i)
            .closest('.connection-info')
            .querySelector('.notification-dot');

        expect(notificationDot).toBeInTheDocument();
    })

    test('displays error message when API call fails', async () => {
        // mock api with error msg
        Api.getConnections.mockRejectedValue(new Error('Failed to fetch connections'));

        render(
            <MemoryRouter>
                <UserProvider
                    currUser={demoUser}>
                    <ConnectedUsers />
                </UserProvider>
            </MemoryRouter>
        );

        // expect error msg to render
        await waitFor(() => {
            const errorMessage = screen.getByText(/Failed to fetch connections/i);
            expect(errorMessage).toBeInTheDocument();
        });
    });
})