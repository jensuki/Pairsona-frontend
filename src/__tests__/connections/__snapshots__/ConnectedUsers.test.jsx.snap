// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ConnectedUsers component > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="ConnectedUsers"
  >
    <h2>
      Your Connections
    </h2>
    <ul
      class="connections-list"
    >
      <li
        class="connection-item"
      >
        <div
          class="connection-info"
        >
          <img
            alt="User Three"
            class="profile-pic"
            loading="lazy"
            src="/assets/default_pic.png"
          />
          <div
            class="user-details"
          >
            <a
              class="user-name"
              data-discover="true"
              href="/profile/user3"
            >
              User Three
            </a>
          </div>
        </div>
        <div
          class="connection-actions"
        >
          <button
            class="chat-btn"
          >
            Chat with User
          </button>
          <button
            class="remove-btn"
          >
            Remove
          </button>
        </div>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
